File: components/deno/Infer.test.ts
import { assertEquals } from "https://deno.land/std@0.137.0/testing/asserts.ts";
import { Constraints } from "./Constraints.ts";
import { inferPattern, inferProgram } from "./Infer.ts";
import { parse, Pattern } from "./Parser.ts";
import {
  createFresh,
  DataDefinition,
  emptyTypeEnv,
  Scheme,
  TArr,
  TCon,
  TVar,
  Type,
  typeBool,
  TypeEnv,
  typeInt,
} from "./Typing.ts";

const assertTypeEquals = (ts: Array<Type>, expected: Array<string>) => {
  assertEquals(ts.map((t) => t.toString()), expected);
};

const assertConstraintsEquals = (
  constraints: Constraints,
  expected: Array<string>,
) => {
  assertEquals(constraints.toString(), expected.join(", "));
};

Deno.test("infer Apply", () => {
  const [constraints, type] = inferProgram(
    emptyTypeEnv
      .extend(
        "a",
        new Scheme(new Set(["T"]), new TArr(new TVar("T"), new TVar("T"))),
      )
      .extend("b", new Scheme(new Set(), typeInt)),
    parse("a b"),
    new Constraints(),
    createFresh(),
  );

  assertConstraintsEquals(constraints, [
    "V1 -> V1 ~ Int -> V2",
  ]);
  assertTypeEquals(type, ["V2"]);
});

Deno.test("infer If", () => {
  const [constraints, type] = inferProgram(
    emptyTypeEnv
      .extend("a", new Scheme(new Set(["S"]), new TVar("S")))
      .extend("b", new Scheme(new Set(), typeInt))
      .extend("c", new Scheme(new Set(["T"]), new TVar("T"))),
    parse("if (a) b else c"),
    new Constraints(),
    createFresh(),
  );

  assertConstraintsEquals(constraints, [
    "V1 ~ Bool",
    "Int ~ V2",
  ]);
  assertTypeEquals(type, ["Int"]);
});

Deno.test("infer Lam", () => {
  const [constraints, type] = inferProgram(
    emptyTypeEnv,
    parse("\\x -> x 10"),
    new Constraints(),
    createFresh(),
  );

  assertConstraintsEquals(constraints, [
    "V1 ~ Int -> V2",
  ]);
  assertTypeEquals(type, ["V1 -> V2"]);
});

Deno.test("infer Let", () => {
  const [constraints, type] = inferProgram(
    emptyTypeEnv,
    parse("let x = 10 and y = x + 1 ; y"),
    new Constraints(),
    createFresh(),
  );

  assertConstraintsEquals(constraints, [
    "Int -> Int -> V1 ~ Int -> Int -> Int",
  ]);
  assertTypeEquals(type, ["(Int * Int)", "Int"]);
});

Deno.test("infer LBool", () => {
  const [constraints, type] = inferProgram(
    emptyTypeEnv,
    parse("True"),
    new Constraints(),
    createFresh(),
  );

  assertConstraintsEquals(constraints, []);
  assertTypeEquals(type, ["Bool"]);
});

Deno.test("infer LInt", () => {
  const [constraints, type] = inferProgram(
    emptyTypeEnv,
    parse("123"),
    new Constraints(),
    createFresh(),
  );

  assertEquals(constraints.constraints.length, 0);
  assertTypeEquals(type, ["Int"]);
});

Deno.test("infer LString", () => {
  const [constraints, type] = inferProgram(
    emptyTypeEnv,
    parse('"hello"'),
    new Constraints(),
    createFresh(),
  );

  assertEquals(constraints.constraints.length, 0);
  assertTypeEquals(type, ["String"]);
});

Deno.test("infer LTuple", () => {
  const [constraints, type] = inferProgram(
    emptyTypeEnv,
    parse('(1, "hello", True)'),
    new Constraints(),
    createFresh(),
  );

  assertEquals(constraints.constraints.length, 0);
  assertTypeEquals(type, ["(Int * String * Bool)"]);
});

Deno.test("infer LUnit", () => {
  const [constraints, type] = inferProgram(
    emptyTypeEnv,
    parse("()"),
    new Constraints(),
    createFresh(),
  );

  assertEquals(constraints.constraints.length, 0);
  assertTypeEquals(type, ["()"]);
});

Deno.test("infer Match", () => {
  const [constraints, type] = inferProgram(
    emptyTypeEnv,
    parse("match (1, 2) with (x, y) -> x + y"),
    new Constraints(),
    createFresh(),
  );

  assertConstraintsEquals(constraints, [
    "V2 -> V3 -> V4 ~ Int -> Int -> Int",
    "(V2 * V3) ~ (Int * Int)",
    "V4 ~ V1",
  ]);
  assertTypeEquals(type, ["V1"]);
});

Deno.test("infer PBool pattern", () => {
  assertInferPattern(
    { type: "PBool", value: true },
    [],
    "Bool",
  );

  assertInferPattern(
    { type: "PBool", value: false },
    [],
    "Bool",
  );
});

Deno.test("infer PCons pattern", () => {
  const origEnv = emptyTypeEnv.addData(
    new DataDefinition("List", ["a"], [
      new TCon("Nil", []),
      new TCon("Cons", [new TVar("a"), new TCon("List", [new TVar("a")])]),
    ]),
  );

  assertInferPatternWithEnv(
    origEnv,
    { type: "PCons", name: "Nil", args: [] },
    [],
    "List V1",
    origEnv,
  );

  assertInferPatternWithEnv(
    origEnv,
    {
      type: "PCons",
      name: "Cons",
      args: [{ type: "PVar", name: "x" }, { type: "PVar", name: "xs" }],
    },
    ["V2 ~ V1", "V3 ~ List V1"],
    "List V1",
    origEnv
      .extend("x", new Scheme(new Set(), new TVar("V2")))
      .extend("xs", new Scheme(new Set(), new TVar("V3"))),
  );
});

Deno.test("infer PInt pattern", () => {
  assertInferPattern(
    { type: "PInt", value: 123 },
    [],
    "Int",
  );
});

Deno.test("infer PString pattern", () => {
  assertInferPattern(
    { type: "PString", value: "hello" },
    [],
    "String",
  );
});

Deno.test("infer PTuple pattern", () => {
  assertInferPattern(
    {
      type: "PTuple",
      values: [
        { type: "PBool", value: true },
        { type: "PInt", value: 123 },
        { type: "PString", value: "hello" },
        { type: "PUnit" },
      ],
    },
    [],
    "(Bool * Int * String * ())",
  );
});

Deno.test("infer PUnit pattern", () => {
  assertInferPattern(
    { type: "PUnit" },
    [],
    "()",
  );
});

Deno.test("infer PVar pattern", () => {
  assertInferPattern(
    { type: "PVar", name: "x" },
    [],
    "V1",
    emptyTypeEnv.extend("x", new Scheme(new Set(), new TVar("V1"))),
  );
});

Deno.test("infer PWildCard pattern", () => {
  assertInferPattern(
    { type: "PWildcard" },
    [],
    "V1",
  );
});

Deno.test("infer Op", () => {
  const scenario = (input: string, resultType: Type) => {
    const [constraints, type] = inferProgram(
      emptyTypeEnv
        .extend("a", new Scheme(new Set(["T"]), new TVar("T")))
        .extend("b", new Scheme(new Set(["T"]), new TVar("T"))),
      parse(input),
      new Constraints(),
      createFresh(),
    );

    assertConstraintsEquals(constraints, [
      `V1 -> V2 -> V3 ~ Int -> Int -> ${resultType}`,
    ]);
    assertTypeEquals(type, ["V3"]);
  };

  scenario("a + b", typeInt);
  scenario("a - b", typeInt);
  scenario("a * b", typeInt);
  scenario("a / b", typeInt);
  scenario("a == b", typeBool);
});

Deno.test("infer Var", () => {
  const [constraints, type] = inferProgram(
    emptyTypeEnv.extend(
      "a",
      new Scheme(new Set(["T"]), new TArr(new TVar("T"), new TVar("T"))),
    ),
    parse("a"),
    new Constraints(),
    createFresh(),
  );

  assertConstraintsEquals(constraints, []);
  assertTypeEquals(type, ["V1 -> V1"]);
});

const assertInferPatternWithEnv = (
  defaultEnv: TypeEnv,
  input: Pattern,
  expectedConstraints: Array<string>,
  expectedType: string,
  expectedTypeEnv: TypeEnv = emptyTypeEnv,
) => {
  const constraints = new Constraints();

  const [type, typeEnv] = inferPattern(
    input,
    defaultEnv,
    constraints,
    createFresh(),
  );

  assertConstraintsEquals(constraints, expectedConstraints);
  assertEquals(type.toString(), expectedType);
  assertEquals(typeEnv, expectedTypeEnv);
};

const assertInferPattern = (
  input: Pattern,
  expectedConstraints: Array<string>,
  expectedType: string,
  expectedTypeEnv: TypeEnv = emptyTypeEnv,
) =>
  assertInferPatternWithEnv(
    emptyTypeEnv,
    input,
    expectedConstraints,
    expectedType,
    expectedTypeEnv,
  );

File: components/kotlin/app/src/test/kotlin/tlca/InferTest.kt
package tlca

import kotlin.test.Test
import kotlin.test.assertEquals

class InferTest {
    @Test
    fun inferApply() {
        assertInference(
            emptyTypeEnv
                    + Pair("a", Scheme(setOf("T"), TArr(TVar("T"), TVar("T"))))
                    + Pair("b", Scheme(emptySet(), typeInt)),
            "a b",
            listOf("V1 -> V1 ~ Int -> V2"),
            listOf("V2")
        )
    }

    @Test
    fun inferIf() {
        assertInference(
            emptyTypeEnv
                    + Pair("a", Scheme(setOf("S"), TVar("S")))
                    + Pair("b", Scheme(emptySet(), typeInt))
                    + Pair("c", Scheme(setOf("S"), TVar("S"))),
            "if (a) b else c",
            listOf(
                "V1 ~ Bool",
                "Int ~ V2"
            ),
            listOf("Int")
        )
    }

    @Test
    fun inferLam() {
        assertInference(
            emptyTypeEnv,
            "\\x -> x 10",
            listOf("V1 ~ Int -> V2"),
            listOf("V1 -> V2")
        )
    }

    @Test
    fun inferLBool() {
        assertInference("True", "Bool")
    }

    @Test
    fun inferLInt() {
        assertInference("123", "Int")
    }

    @Test
    fun inferLString() {
        assertInference("\"hello\"", "String")
    }

    @Test
    fun inferLTuple() {
        assertInference("(1, \"hello\", (), True)", "(Int * String * () * Bool)")
    }

    @Test
    fun inferLet() {
        assertInference(
            emptyTypeEnv,
            "let x = 10 and y = x + 1 ; y",
            listOf(
                "Int -> Int -> V1 ~ Int -> Int -> Int"
            ),
            listOf("(Int * Int)", "Int")
        )
    }

    @Test
    fun inferMatch() {
        assertInference(
            emptyTypeEnv,
            "match (1, 2) with (x, y) -> x + y",
            listOf(
                "V2 -> V3 -> V4 ~ Int -> Int -> Int",
                "(V2 * V3) ~ (Int * Int)",
                "V4 ~ V1",
            ),
            listOf("V1")
        )
    }

    @Test
    fun inferOp() {
        fun scenario(input: String, resultType: Type) {
            assertInference(
                emptyTypeEnv
                        + Pair("a", Scheme(setOf("T"), TVar("T")))
                        + Pair("b", Scheme(setOf("T"), TVar("T"))),
                input,
                listOf(
                    "V1 -> V2 -> V3 ~ Int -> Int -> $resultType"
                ),
                listOf("V3")
            )
        }

        scenario("a + b", typeInt)
        scenario("a - b", typeInt)
        scenario("a * b", typeInt)
        scenario("a / b", typeInt)
        scenario("a == b", typeBool)
    }

    @Test
    fun inferVar() {
        assertInference(
            emptyTypeEnv + Pair("a", Scheme(setOf("T"), TArr(TVar("T"), TVar("T")))),
            "a",
            emptyList(),
            listOf("V1 -> V1")
        )
    }

    @Test
    fun inferBoolPattern() {
        assertInferPattern(PBoolPattern(true), emptyList(), "Bool")
    }

    @Test
    fun inferConsPattern() {
        val origEnv = emptyTypeEnv.addData(
            DataDefinition(
                "List", listOf("a"), listOf(
                    DataConstructor("Nil", emptyList()),
                    DataConstructor("Cons", listOf(TVar("a"), TArr(TCon("List"), TVar("a"))))
                )
            )
        )

        assertInferPatternWithEnv(origEnv, PDataPattern("Nil", emptyList()), emptyList(), "List V1", origEnv)
        assertInferPatternWithEnv(
            origEnv,
            PDataPattern("Cons", listOf(PVarPattern("x"), PVarPattern("xs"))),
            listOf("V2 ~ V1, V3 ~ List -> V1"),
            "List V1",
            origEnv
                .extend("x", Scheme(emptySet(), TVar("V2")))
                .extend("xs", Scheme(emptySet(), TVar("V3")))
        )
    }

    @Test
    fun inferIntPattern() {
        assertInferPattern(PIntPattern(123), emptyList(), "Int")
    }

    @Test
    fun inferStringPattern() {
        assertInferPattern(PStringPattern("Hello"), emptyList(), "String")
    }

    @Test
    fun inferTuplePattern() {
        assertInferPattern(
            PTuplePattern(
                listOf(
                    PBoolPattern(true),
                    PIntPattern(123),
                    PStringPattern("Hello"),
                    PUnitPattern,
                )
            ), emptyList(), "(Bool * Int * String * ())"
        )
    }

    @Test
    fun inferUnitPattern() {
        assertInferPattern(PUnitPattern, emptyList(), "()")
    }

    @Test
    fun inferVarPattern() {
        assertInferPattern(PVarPattern("x"), emptyList(), "V1", emptyTypeEnv.extend("x", Scheme(emptySet(), TVar("V1"))))
    }

    @Test
    fun inferWildcardPattern() {
        assertInferPattern(PWildcardPattern, emptyList(), "V1", emptyTypeEnv)
    }

    private fun assertInference(typeEnv: TypeEnv, input: String, expectedConstraints: List<String>, expectedTypes: List<String>) {
        val inferResult = infer(typeEnv, parseExpressions(input), Constraints(), Pump())

        assertConstraints(inferResult.constraints, expectedConstraints)
        assertEquals(expectedTypes, inferResult.types.map { it.toString() })
    }

    private fun assertInference(input: String, expectedType: String) {
        assertInference(emptyTypeEnv, input, emptyList(), listOf(expectedType))
    }

    private fun assertConstraints(constraints: Constraints, expected: List<String>) {
        assertEquals(expected.joinToString(", "), constraints.toString())
    }

    private fun assertInferPatternWithEnv(
        defaultEnv: TypeEnv,
        pattern: Pattern,
        expectedConstraints: List<String> = emptyList(),
        expectedType: String,
        expectedTypeEnv: TypeEnv = emptyTypeEnv
    ) {
        val constraints = Constraints()

        val (type, typeEnv) = inferPattern(pattern, defaultEnv,     constraints, Pump())

        assertConstraints(constraints, expectedConstraints)
        assertEquals(type.toString(), expectedType)
        assertEquals(typeEnv, expectedTypeEnv)
    }

    private fun assertInferPattern(
        pattern: Pattern,
        expectedConstraints: List<String> = emptyList(),
        expectedType: String,
        expectedTypeEnv: TypeEnv = emptyTypeEnv
    ) = assertInferPatternWithEnv(emptyTypeEnv, pattern, expectedConstraints, expectedType, expectedTypeEnv)
}
